RockLib.Configuration.ObjectFactory/README.md

```diff
@@ -0,0 +1,392 @@
+
+```json
+{
+    "bar": 123,
+    "baz": {
+        "type": "MyNamespace.BazDerived, MyAssembly",
+        "value": {
+            "qux": "abc",
+            "corge": true
+        }
+    }
+}
+```
+
+Being able to specify a type means that interfaces and abstract classes are supported.
+
+```c#
+public class Foo
+{
+    public IBar Bar { get; set; }
+    public BazBase Baz { get; set; }
+}
+
+public interface IBar
+{
+}
+
+public class Bar : IBar
+{
+    public int Qux { get; set; }
+}
+
+public abstract class BazBase
+{
+}
+
+public class Baz : BazBase
+{
+    public bool Corge { get; set; }
+}
+```
+
+is mappable from:
+
+```json
+{
+    "bar": {
+        "type": "MyNamespace.Bar, MyAssembly",
+        "value": {
+            "qux": "abc"
+        }
+    },
+    "baz": {
+        "type": "MyNamespace.Baz, MyAssembly",
+        "value": {
+            "corge": true
+        }
+    }
+}
+```
+
+### Lists
+
+RockLib.Configuration.ObjectFactory currently supports the following list types:
+
+- Any single-dimension array
+- `List<T>`
+- `IList<T>`
+- `ICollection<T>`
+- `IEnumerable<T>`
+- `IReadOnlyCollection<T>`
+- `IReadOnlyList<T>`
+
+If a list property is readonly, one of the following conditions must be met:
+
+- The property type is one of the following:
+  - `List<T>`
+  - `IList<T>`
+  - `ICollectionList<T>`
+- The property type implements `IList` and has a single `Add` method that has one parameter with a type other than `object`.
+
+If a class has a list-type property, but the application only want to define one item, the configuration can be flattened. For the `Foo` class:
+
+```c#
+public class Foo
+{
+    public IEnumerable<Bar> Bars { get; set; }
+}
+
+public class Bar
+{
+    public string Baz { get; set; }
+    public int Qux { get; set; }
+}
+```
+
+instead of a configuration like this:
+
+```json
+{
+    "bars": [
+        {
+            "baz": "abc",
+            "qux": 123
+        }
+    ]
+}
+```
+
+can be rewritten like this:
+
+```json
+{
+    "bars": {
+        "baz": "abc",
+        "qux": 123
+    }
+}
+```
+
+### Default Types
+
+Given a type `Foo` that has an abstract property `Bar`:
+
+```c#
+public class Foo
+{
+    public IBar Bar { get; set; }
+}
+
+public interface IBar
+{
+}
+
+public class DefaultBar : IBar
+{
+    public int Baz { get; set; }
+}
+
+We would like to be able to define the `DefaultBar` type as the default type - if the type is not otherwise specified, create an instance of `DefaultBar`. There are two ways of specifying the default type: by property or by type. When specified by property, we want to say: "when setting the `Foo.Bar` property, if the configuration is *not* type-specified, set the property to an instance of `DefaultBar`." When specified by type we want to say: "whenever creating an instance of `IBar`, if the configuration is *not* type-specified, create an instance of `DefaultBar`.
```

Comment:
Should this close off the code section?  It seems to look funny here in the readme